src/soc/rockchip/rk3399.c

```diff
@@ -0,0 +1,538 @@
+	{"GPIO4_D5", 4, {CRU_CLKGATE_CON31, 5}, {GRF_GPIO4D_IOMUX, 10}, {GPIO_SWPORTA_DDR, 29}, {GPIO_SWPORTA_DR, 29}, FUNCTION_DIGITAL, PINMODE_NOT_SET, 0},
+	{"GPIO4_D6", 4, {CRU_CLKGATE_CON31, 5}, {GRF_GPIO4D_IOMUX, 12}, {GPIO_SWPORTA_DDR, 30}, {GPIO_SWPORTA_DR, 30}, FUNCTION_DIGITAL, PINMODE_NOT_SET, 0},
+	{"GPIO4_D7", 4, {CRU_CLKGATE_CON31, 5}, {GRF_GPIO4D_IOMUX, 14}, {GPIO_SWPORTA_DDR, 31}, {GPIO_SWPORTA_DR, 31}, FUNCTION_DIGITAL, PINMODE_NOT_SET, 0},
+};
+
+static int rk3399Setup(void) {
+	if((rk3399->fd = open("/dev/mem", O_RDWR | O_SYNC)) < 0) {
+		wiringXLog(LOG_ERR, "wiringX failed to open /dev/mem for raw memory access");
+		return -1;
+	}
+	for(int i = 0; i < GPIO_BANK_COUNT; i++) {
+		if((rk3399->gpio[i] = (unsigned char *)mmap(0, rk3399->page_size, PROT_READ | PROT_WRITE, MAP_SHARED, rk3399->fd, rk3399->base_addr[i])) == NULL) {
+			wiringXLog(LOG_ERR, "wiringX failed to map The %s %s GPIO memory address", rk3399->brand, rk3399->chip);
+			return -1;
+		}
+	}
+	if((cru_register_virtual_address = (unsigned char *)mmap(0, rk3399->page_size, PROT_READ | PROT_WRITE, MAP_SHARED, rk3399->fd, CRU_REGISTER_PHYSICAL_ADDRESS)) == NULL) {
+		wiringXLog(LOG_ERR, "wiringX failed to map The %s %s CRU memory address", rk3399->brand, rk3399->chip);
+		return -1;
+	}
+	if((pmugrf_register_virtual_address = (unsigned char *)mmap(0, rk3399->page_size, PROT_READ | PROT_WRITE, MAP_SHARED, rk3399->fd, PMUGRF_REGISTER_PHYSICAL_ADDRESS)) == NULL) {
+		wiringXLog(LOG_ERR, "wiringX failed to map The %s %s PMUGRF memory address", rk3399->brand, rk3399->chip);
+		return -1;
+	}
+	if((grf_register_virtual_address = (unsigned char *)mmap(0, rk3399->page_size, PROT_READ | PROT_WRITE, MAP_SHARED, rk3399->fd, GRF_REGISTER_PHYSICAL_ADDRESS)) == NULL) {
+		wiringXLog(LOG_ERR, "wiringX failed to map The %s %s GRF memory address", rk3399->brand, rk3399->chip);
+		return -1;
+	}
+
+	return 0;
+}
+
+static char *rk3399GetPinName(int pin) {
+	return rk3399->layout[pin].name;
+}
+
+static void rk3399SetMap(int *map, size_t size) {
+	rk3399->map = map;
+	rk3399->map_size = size;
+}
+
+static void rk3399SetIRQ(int *irq, size_t size) {
+	rk3399->irq = irq;
+	rk3399->irq_size = size;
+}
+
+struct layout_t * rk3399GetLayout(int i, int* mapping) {
+	struct layout_t *pin = NULL;
+	unsigned int *grf_reg = NULL;
+	unsigned int iomux_value = 0;
+	if(mapping == NULL) {
+		wiringXLog(LOG_ERR, "The %s %s has not yet been mapped", rk3399->brand, rk3399->chip);
+		return NULL;
+	}
+	if(wiringXValidGPIO(i) != 0) {
+		wiringXLog(LOG_ERR, "The %i is not the right GPIO number");
+		return NULL;
+	}
+	if(rk3399->fd <= 0 || rk3399->gpio == NULL) {
+		wiringXLog(LOG_ERR, "The %s %s has not yet been setup by wiringX", rk3399->brand, rk3399->chip);
+		return NULL;
+	}
+
+	pin = &rk3399->layout[mapping[i]];
+	if(pin->bank < 0 || pin->bank >= GPIO_BANK_COUNT) {
+		wiringXLog(LOG_ERR,"pin->bank out of range: %i, expect 0~4", pin->bank);
```

Comment:
Spaces:
`wiringXLog(LOG_ERR, "pin->bank out of range: %i, expect 0~4", pin->bank);`